### PR TITLE
Add presubmit check for new .cc/.c/.mm/.m source files without corresponding GN changes

### PR DESCRIPTION
**Pull request by yrliou**
Resolves https://github.com/brave/brave-browser/issues/35678

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: PRESUBMIT.py [hunk 1]
```diff
@@ -280,6 +280,48 @@ def CheckLicense(input_api, output_api):
...
+
+    problems = []
+    for source in new_sources:
+        basename = input_api.os_path.basename(source)
+        if basename not in all_gn_changed_contents:
+            problems.append(source)
+
+    if problems:
+        return [
+            output_api.PresubmitError(
```
Comment (yrliou) [inline PRESUBMIT.py:316]: Let me know if error is desired or warning only.

File: PRESUBMIT.py [hunk 2]
```diff
@@ -280,6 +280,48 @@ def CheckLicense(input_api, output_api):
     return result
 
 
+def CheckNewSourceFileWithoutGnChangeOnUpload(input_api, output_api):
+    """Checks newly added .cc/.h/.mm files have corresponding GN changes."""
+    files_to_check = (
+        r'.+\.cc$',
+        r'.+\.h$',
```
Comment (goodov) [inline PRESUBMIT.py:287]: `.h` files already validated by the existing upstream check (we do run it). Let's validate only `.cc, .c, .mm, .m` in our check.

File: PRESUBMIT.py [hunk 3]
```diff
@@ -280,6 +280,48 @@ def CheckLicense(input_api, output_api):
...
+        return input_api.FilterSourceFile(f, files_to_check=files_to_check)
+
+    new_sources = []
+    for f in input_api.AffectedSourceFiles(new_sources):
+        if f.Action() != 'A':
+            continue
+        new_sources.append(f.LocalPath())
+
+    def gn_files(f):
+        return input_api.FilterSourceFile(f, files_to_check=(r'.+\.gn', ))
```
Comment (goodov) [inline PRESUBMIT.py:301]: this should exclude `chromium_src` dir.

Comment (goodov) [inline PRESUBMIT.py:301]: can you also test that this covers `.gn` and `.gni` files.

Comment (yrliou) [reply PRESUBMIT.py]: I think it was but might involve other stuff by accident, changed to .+\.gn$ and .+\.gni$

Comment (yrliou) [reply PRESUBMIT.py]: Now skipping both for new sources and gn_files.
I believe most of the time new source in chromium_src won't have BUILD.gn changes.
There's a case (maybe just a corner case?) where we might add our own test file that's not for override under chromium_src folder and will need gn changes tho, this PR's current state does not catch that.

**Review by goodov — APPROVED** [review]
thanks!

**Pull request merged**